app/src/main/java/com/hover/stax/database/DatabaseRepo.kt

```diff
@@ -150,17 +152,22 @@ class DatabaseRepo(db: AppDatabase, sdkDb: HoverRoomDatabase) {
                 val a = getAction(intent.getStringExtra(HoverAction.ID_KEY))
                 val channel = getChannel(a.channel_id)
                 val contact = StaxContact.findOrInit(intent, channel.countryAlpha2, t, this)
+                var isNew = false
 
                 if (contact.accountNumber != null)
                     save(contact)
 
                 if (t == null) {
+                    isNew = true
                     c?.let { Utils.logAnalyticsEvent(c.getString(R.string.initializing_ussd_services), c) }
                     t = StaxTransaction(intent, a, contact, c)
                     transactionDao.insert(t)
                     t = transactionDao.getTransaction(t.uuid)
                 }
+
                 t!!.update(intent, a, contact, c)
+                if( !isNew && shouldFlagFailure(t, c)) t.setFailed_Incomplete()
```

Comment:
Instead of doing all this work here, just add the logic for this directly to the update method of the staxtransaction object. That way the DatabaseRepo doesn't need to know about the internals of transactions.

Reply:
Thanks, changes made